### ISSUE TITLE
Increasing overhead untill initial data fetch using make_reader

### ISSUE THREAD

**Issue by GregAru** [uber/petastorm#443]
I have a ~500GB dataset residing in S3 generated using `materialize_dataset`.
The number of parquet files in this dataset is 3678. Each file size is ~135MB.
The dataset is partitioned and each partition holds between 5 to 10 parquet files.

When using `make_reader` on this dataset with `rowgroup_selector` applied I need to wait 2,3 minutes until the data is fetched. From a short debugging I can see that initially the process gets stuck at: `dataset_metadata.get_schema_from_dataset_url(dataset_url, hdfs_driver=hdfs_driver)` (line 132) in `reader.py`

Perhaps I am wrong but it looks as if `ParquetDataset` scans all the parquet files and not just the _common_metadata. Is this an expected behavior? If so this is worrying - since I intend to increase the dataset size to at least 10PB

Additional information:
The _common_metadata file size of the dataset is ~500KB and there is no _metadata file.

**Comment by GregAru** [comment]
Perhaps I misunderstand but in _common_metadata under dataset-toolkit.num_row_groups_per_file.v1
I can essentially see the full manifest of my dataset. Not sure why the full traversal is needed.

**Comment by GregAru** [comment]
@selitvin , Any suggestions?

**Comment by selitvin** [comment]
Hi. Thanks for the ping - somehow missed this issue. 
 think you are right. We can restructure `get_schema_from_dataset_url` only to load manifest and not create `ParquetDataset` instance that traverses the directory. We also do it again, down the road, in the `Reader` class implementation. 

I see that `ParquetDataset` takes `metadata_nthreads` argument. I would assume that, at least as a temporary relief we can try passing a non default > 1, which should speed up the directory traversal significantly. 

Long term, I can try remove the dependency on `pq.ParquetDataset` and work off metadata information only. Does this makes sense?

**Comment by cupdike** [comment]
I'm no expert, but from what I've read, _common_metadata, is a parquet-mr construct.  And it is deprecated and to be eliminated in parquet-mr 2.0.0:  https://github.com/apache/parquet-mr/blob/parquet-1.10.x/parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ParquetFileWriter.java#L752

**Comment by selitvin** [comment]
Yep. Heard about it as well. Maybe we should start storing our data in a custom files then? I remember there was an issue with making pyarrow ignore `_*` files in parquet directories, but maybe there is a solution for that.

**Comment by GregAru** [comment]
Makes absolute sense.
I think that `metadata_nthreads` should be exposed in `make_reader`.
I have played with it a bit - it does improve performance but still it is ineffective for the file amounts I will have on S3. Using _common_metadata only would be the way to go.
